pkg/operator/watchdog/cmd.go

```diff
@@ -0,0 +1,247 @@
+
+	return nil
+}
+
+func (o *FileWatcherOptions) Validate() error {
+	if len(o.ProcessName) == 0 {
+		return fmt.Errorf("process name must be specified")
+	}
+	if len(o.Files) == 0 {
+		return fmt.Errorf("at least one file to observe must be specified")
+	}
+	if len(o.Namespace) == 0 && len(os.Getenv("POD_NAMESPACE")) == 0 {
+		return fmt.Errorf("either namespace flag or POD_NAMESPACE environment variable must be specified")
+	}
+	return nil
+}
+
+// startPidObserver runs a loop that observe changes to the PID of the process we send signals after change is detected.
+// When the PID is observed for the first time, the pidObservedCh is closed.
+// When the PID change is observed (after we got initial PID), the shutdown is called (to signal watchdog to restart).
+func (o *FileWatcherOptions) startPidObserver(shutdown context.CancelFunc, pidObservedCh chan int) {
+	defer shutdown()
+	defer close(pidObservedCh)
+	currentPID := 0
+	for {
+		// get the monitored process PID
+		observedPID, found, err := o.getPIDByNameFn(o.ProcessName)
+		if !found {
+			klog.Warningf("Process %q does not have PID", o.ProcessName)
+			return
+		}
+		if err != nil {
+			klog.Warning(err)
+			return
+		}
+
+		if currentPID == 0 {
+			currentPID = observedPID
+			pidObservedCh <- observedPID
+		}
+
+		// when the PID change, signal shutdown
+		if currentPID != observedPID {
+			return
+		}
+
+		time.Sleep(1 * time.Second)
+	}
+}
+
+// readInitialFileContent reads the content of files specified.
+// This is needed by file observer.
+func (o *FileWatcherOptions) readInitialFileContent() error {
+	o.initialContent = map[string][]byte{}
+	for _, name := range o.Files {
+		content, err := ioutil.ReadFile(name)
+		if err != nil {
+			return err
+		}
+		o.initialContent[name] = content
+	}
+	return nil
+}
+
+// setFilesPathPrefix mutates the file list and prefix every file with /proc/PID/root.
+func setFilesPathPrefix(oldFiles []string, pid int) []string {
+	files := []string{}
+	for _, file := range oldFiles {
+		files = append(files, filepath.Join("/proc", fmt.Sprintf("%d", pid), "root", file))
+	}
+	return files
+}
+
+func (o *FileWatcherOptions) Run(ctx context.Context) error {
+	runCtx, shutdown := context.WithCancel(ctx)
+
+	pidObservedCh := make(chan int)
+	go o.startPidObserver(shutdown, pidObservedCh)
+
+	// Wait while we get the initial PID for the process
+	klog.Infof("Waiting for process %q PID ...", o.ProcessName)
+	currentPID := <-pidObservedCh
+
+	// Mutate path for specified files as '/proc/PID/root/<original path>'
+	// This means side-car container don't have to duplicate the mounts from main container.
+	// This require shared PID namespace feature.
+	o.Files = o.setFilesPathPrefixFn(o.Files, currentPID)
+
+	// Read initial file content. If shared PID namespace does not work, this will error.
+	if err := o.readInitialFileContent(); err != nil {
+		o.recorder.Warningf("FileChangeWatchdogFailed", "Reading initial file content failed: %v", err)
```

Comment:
does this really need events? It's a misconfiguration. I think logs are enough.

Reply:
if we restart/crash the log is gone, don't we learned our lessons ? :-)